File: AIペアプログラミング実践編_ソースコード/practical-csharp/session-5/Ans/Inventory.Api/Services/OrderService.cs
namespace Inventory.Api.Services;

using Inventory.Api.Dtos;
using Inventory.Api.Entities;
using Inventory.Api.Repositories;

public class OrderService
{
    private readonly IOrderRepository _orderRepository;
    private readonly InventoryService _inventoryService;

    public OrderService(IOrderRepository orderRepository, InventoryService inventoryService)
    {
        _orderRepository = orderRepository;
        _inventoryService = inventoryService;
    }

    public Order Create(OrderInput input)
    {
        var order = new Order
        {
            ProductId = input.ProductId,
            Quantity = input.Quantity,
            Status = OrderStatus.Pending,
            CreatedAt = DateTime.Now
        };
        return _orderRepository.Save(order);
    }

    public Order Order(int orderId)
    {
        var order = _orderRepository.FindById(orderId)
            ?? throw new ArgumentException("発注が見つかりません");

        if (order.Status != OrderStatus.Pending)
        {
            throw new InvalidOperationException("発注中の発注のみ発注済みにできます");
        }

        order.Status = OrderStatus.Ordered;
        order.OrderedAt = DateTime.Now;
        return _orderRepository.Save(order);
    }

    public Order Receive(int orderId)
    {
        var order = _orderRepository.FindById(orderId)
            ?? throw new ArgumentException("発注が見つかりません");

        if (order.Status != OrderStatus.Ordered)
        {
            throw new InvalidOperationException("発注済みの発注のみ入荷処理できます");
        }

        // 在庫を更新
        _inventoryService.UpdateQuantity(order.ProductId, TransactionType.In, order.Quantity);

        order.Status = OrderStatus.Received;
        order.ReceivedAt = DateTime.Now;
        return _orderRepository.Save(order);
    }

    public Order Cancel(int orderId)
    {
        var order = _orderRepository.FindById(orderId)
            ?? throw new ArgumentException("発注が見つかりません");

        if (order.Status == OrderStatus.Received)
        {
            throw new InvalidOperationException("入荷済みの発注はキャンセルできません");
        }

        order.Status = OrderStatus.Cancelled;
        return _orderRepository.Save(order);
    }

    public List<Order> FindAll()
    {
        return _orderRepository.FindAll();
    }

    public List<Order> FindByStatus(OrderStatus status)
    {
        return _orderRepository.FindByStatus(status);
    }
}

File: AIペアプログラミング実践編_ソースコード/practical-python/session-5/Ans/inventory/services/order_service.py
"""発注サービス"""
from datetime import datetime

from inventory.models.order import Order, OrderStatus
from inventory.models.transaction import TransactionType

from .dto import OrderInput


class OrderService:
    """発注のビジネスロジック"""

    def __init__(self, repository=None, inventory_service=None):
        self.repository = repository
        self.inventory_service = inventory_service

    def create(self, input_data: OrderInput) -> Order:
        """発注を作成"""
        order = Order(
            product_id=input_data.product_id,
            quantity=input_data.quantity,
            status=OrderStatus.PENDING
        )

        if self.repository:
            return self.repository.save(order)
        else:
            order.save()
            return order

    def order(self, order_id: int) -> Order:
        """発注を発注済みに更新"""
        order = self._get_order(order_id)

        if order.status != OrderStatus.PENDING:
            raise ValueError("発注中の発注のみ発注済みにできます")

        order.status = OrderStatus.ORDERED
        order.ordered_at = datetime.now()

        if self.repository:
            return self.repository.save(order)
        else:
            order.save()
            return order

    def receive(self, order_id: int) -> Order:
        """入荷処理"""
        order = self._get_order(order_id)

        if order.status != OrderStatus.ORDERED:
            raise ValueError("発注済みの発注のみ入荷処理できます")

        # 在庫を更新
        if self.inventory_service:
            self.inventory_service.update_quantity(
                product_id=order.product_id,
                transaction_type=TransactionType.IN,
                quantity=order.quantity
            )

        order.status = OrderStatus.RECEIVED
        order.received_at = datetime.now()

        if self.repository:
            return self.repository.save(order)
        else:
            order.save()
            return order

    def cancel(self, order_id: int) -> Order:
        """発注をキャンセル"""
        order = self._get_order(order_id)

        if order.status == OrderStatus.RECEIVED:
            raise ValueError("入荷済みの発注はキャンセルできません")

        order.status = OrderStatus.CANCELLED

        if self.repository:
            return self.repository.save(order)
        else:
            order.save()
            return order

    def find_all(self):
        """全発注を取得"""
        if self.repository:
            return self.repository.find_all()
        return list(Order.objects.all())

    def find_by_status(self, status: OrderStatus):
        """ステータスで絞り込んで発注を取得"""
        if self.repository:
            return self.repository.find_by_status(status)
        return list(Order.objects.filter(status=status))

    def _get_order(self, order_id: int) -> Order:
        """発注を取得"""
        if self.repository:
            order = self.repository.find_by_id(order_id)
        else:
            order = Order.objects.filter(id=order_id).first()

        if order is None:
            raise ValueError("発注が見つかりません")

        return order
